### ISSUE TITLE
Random 0 data receiving 

### ISSUE THREAD

**Issue by shridattdudhat** [luksal/ESP32-DMX-RX#19]
Hi @luksal , I tested your example, but it receives 0 randomly, I cross checked the data using logic analyser, the data is perfect but esp32 receives 0. 
![WhatsApp Image 2021-09-15 at 3 28 57 PM](https://user-images.githubusercontent.com/28555587/133928287-1f471bb0-d6b6-482b-ace8-7f5ce5b730e5.jpeg)

**Comment by yngndrw** [comment]
I've noticed this as well.

It appears that `event.size` from the uart receive queue is sometimes smaller than expected. (https://github.com/luksal/ESP32-DMX/blob/master/src/dmx.cpp#L252)

My work-around for this was to pass the number of expected DMX channels into the `DMX::Initialize()` method and verifying `event.size` against this (`event.size` should be one more than the expected number of channels) before copying the data into `dmx_data`.

**Comment by luksal** [comment]
> My work-around for this was to pass the number of expected DMX channels into the `DMX::Initialize()` method

But how you can know, how many channels were transmitted?

**Comment by macdroid53** [comment]
The actual name of the protocol is DMX512.

It is intended to transmit 512 values per universe.

Any device that does not transmit 512 channels is using a subset of the
specification.

There is no ID of the channel in the transmission, other than the count of
bytes. So, a receiver may choose to only receive a set of bytes (by
counting received bytes and pickling which to use based on a setting in the
receiver). But, the is no way for a sending device to specify to receivers
which of the 512 it is sending.

Most professional devices are receiving or sending 512. The sending and
receiving of less than 512 has been added by simple implementations in DIY
settings where the DIY'er is in control of both the send and receive design.

**Comment by yngndrw** [comment]
> > My work-around for this was to pass the number of expected DMX channels into the `DMX::Initialize()` method
> 
> But how you can know, how many channels were transmitted?

There are a couple of options depending on your scenario.

The first is where you know the device which will be sending data. In my case I'm making a set of room lights which will be controlled by an in-wall dimmer, so I know exactly what device will be sending the data and exactly how many slots it will be sending.

The other option would be to auto-detect the number of slots. You could take a sample of let's say 10 packets and if they all have the same number of slots you can use that value until you next restart. If they don't all agree you can take another sample until you have a valid set of data to work from.


> The actual name of the protocol is DMX512. It is intended to transmit 512 values per universe. Any device that does not transmit 512 channels is using a subset of the specification. There is no ID of the channel in the transmission, other than the count of bytes. So, a receiver may choose to only receive a set of bytes (by counting received bytes and pickling which to use based on a setting in the receiver). But, the is no way for a sending device to specify to receivers which of the 512 it is sending. Most professional devices are receiving or sending 512. The sending and receiving of less than 512 has been added by simple implementations in DIY settings where the DIY'er is in control of both the send and receive design.

I have to disagree with this. The closest I've found to a specification for DMX512 is this: (If there's a better or more current specification, please let me know.)
https://tsp.esta.org/tsp/documents/docs/ANSI-ESTA_E1-11_2008R2018.pdf

Section 8.1 on page 13 mentions "_up to_":
> DMX512 slots shall be transmitted sequentially in asynchronous serial format, beginning with slot 0 and ending
with the last implemented slot, **up to slot 512** (a maximum total of 513 slots).

Later on, section 8.7 on page 16 further clarifies:
> **8.7 Minimum number of data slots**
There shall be **no minimum number of data slots** on the data link. DMX512 data packets with **fewer than 512
slots may be transmitted**, subject to the minimum timing requirements of this Standard – see clause 8.10 and
figure 5.

There is no ambiguity in the addressing, as the slots always start from 1.

Of course, we all use 3 pin XLR connectors which are explicitly forbidden by the specification, so perhaps following the specification is rather moot at this point. (Sections 7.1.1 and 7.1.2 on page 12.)

As an aside, section 4.6 on page 6 mentions that it is _preferred_ that transmitting devices are earth ground referenced and that receiving devices are isolated.

**Comment by macdroid53** [comment]
Easy to interpret the spec as you wish, if, you are in control of what is
being transmitted.

If you expect your receiver to be used in a setting, outside of your
control, possibly, professional, you don't control what is sent. The
receiving device needs to be able to receive all 512, be addressable to any
of the 512 (typically, in sequential groups of 4, anywhere in the 512) and
probably decide the universe.

**Comment by luksal** [comment]
> But how you can know, how many channels were transmitted?

To clarify my statement:

The interrupt is NOT triggered when a complete DMX 512 packet has been received. But if the internal UART buffer is full or another UART event (break etc.) is present.

Therefore `event.size` should contain the amount of data in the buffer. However, there is a bug within the UART implementation in the ESP-IDF, whereby `event.size` always has the same size (I seem to remember that it was 48) see also issue #1 .

So yes, it can help to specify how many channels you expect, however the simple comparison only works for less than 48 channels. For anything above that, you have to validate where in the data stream you are already.

In summary, there are massive problems and open bugs in the UART implementation of the ESP-IDF. Especially when it comes to time critical or protocols that use special features of the UART specification. And as long as these are not fixed, I think it makes more sense to write a low-level driver yourself instead of relying on the ESP-IDF implementation.

Unfortunately, I don't have the time or need to do this, as the code has worked great in the one project I use it in with various DMX interfaces, consoles and ArtNet nodes. However, I also only use PWM and SPI peripherals in conjunction with DMX, no WLAN or Bluetooth connections. I can imagine that this could contribute to the misbehavior of the chip.

**Comment by yngndrw** [comment]
> Easy to interpret the spec as you wish, if, you are in control of what is being transmitted. If you expect your receiver to be used in a setting, outside of your control, possibly, professional, you don't control what is sent. The receiving device needs to be able to receive all 512, be addressable to any of the 512 (typically, in sequential groups of 4, anywhere in the 512) and probably decide the universe.

Well the simple answer for this is to automatically detect the number of slots as that should be consistent.

Section 8.5.1 on page 14 covers this:
> Once a controller is configured for a particular application, all NULL START Code packets should have the same number of slots.

It's perfectly reasonable for a device which doesn't follow the specification to not work correctly, but all of the devices discussed so far do infact follow the specification as does my suggested work-around.

> > But how you can know, how many channels were transmitted?
> 
> To clarify my statement:
> 
> The interrupt is NOT triggered when a complete DMX 512 packet has been received. But if the internal UART buffer is full or another UART event (break etc.) is present.
> 
> Therefore `event.size` should contain the amount of data in the buffer. However, there is a bug within the UART implementation in the ESP-IDF, whereby `event.size` always has the same size (I seem to remember that it was 48) see also issue #1 .
> 
> So yes, it can help to specify how many channels you expect, however the simple comparison only works for less than 48 channels. For anything above that, you have to validate where in the data stream you are already.
> 
> In summary, there are massive problems and open bugs in the UART implementation of the ESP-IDF. Especially when it comes to time critical or protocols that use special features of the UART specification. And as long as these are not fixed, I think it makes more sense to write a low-level driver yourself instead of relying on the ESP-IDF implementation.
> 
> Unfortunately, I don't have the time or need to do this, as the code has worked great in the one project I use it in with various DMX interfaces, consoles and ArtNet nodes. However, I also only use PWM and SPI peripherals in conjunction with DMX, no WLAN or Bluetooth connections. I can imagine that this could contribute to the misbehavior of the chip.

Ah, my situation was for a device which sent 12 slots so I didn't experience the issue with more than 48 slots. In my case, `event.size` would be 13 for most of the time, except for the errors where I'd see an `event.size` of 3-4 with data from the end of a packet. (With and without the semaphores - As an aside `#define DMX_IGNORE_THREADSAFETY 0` still defines `DMX_IGNORE_THREADSAFETY` which means that the semaphores are disabled by default even with the setting of `0`)

For the case where there's more than 48 channels and the data is split over multiple events, I'd suggest building up two buffers. The first would be filled by the event data, moving on to the next set of 48 slots each time. When the next break is seen, this buffer can be verified (Keep track of the total number of slots stored in the preceding packet and compare that to the number expected) and if it is valid, it can be copied into the second buffer which is exposed to the user of the library. (The copy between the first buffer and the second buffer should be protected by a semaphore)

This means that data is delayed by one packet and that invalid packets are discarded, but the specification does allow for some packets to be discarded. Section 8.5.1 on page 14:
> There is no guarantee that all NULL START Code packets will be delivered to all devices. Data sent using NULL
START Codes should be of a type where loss of packet does not greatly affect the operation of the device.
Hence data sent should be of the current value of a parameter, not a command to execute a routine

**Comment by luksal** [comment]
>As an aside `#define DMX_IGNORE_THREADSAFETY 0` still defines `DMX_IGNORE_THREADSAFETY` which means that the semaphores are disabled by default even with the setting of `0`)

Shame on me: that's correct and a stupid careless mistake when copying the code from your own build environment back to this arduinuisch library.

**Comment by luksal** [comment]
>For the case where there's more than 48 channels and the data is split over multiple events, I'd suggest building up two buffers. The first would be filled by the event data, moving on to the next set of 48 slots each time. When the next break is seen, this buffer can be verified (Keep track of the total number of slots stored in the preceding packet and compare that to the number expected) and if it is valid, it can be copied into the second buffer which is exposed to the user of the library. (The copy between the first buffer and the second buffer should be protected by a semaphore)

This is a valid solution for the invalid data problem, but still won't fix #1 where we are unable to receive more than 480 channels, because 480+48 would be 528 which is after the BREAK event, which clears the receive buffer...

**Comment by yngndrw** [comment]
I'm afraid I don't have a better suggestion for the issue in #1 aside from what has already been mentioned within the thread you have created in `espressif/esp-idf`.

**Comment by macdroid53** [comment]
I have not actually looked at this for a long time. My customer lost
interest in the project, so I put it on the shelf.

If I remember correctly, there was a buffer size of 400-ish and when the
low level driver saw the break at the beginning of the transmission it
basically lost the remaining channels up to 512. I think I experimented a
bit trying to get the remainder and was not successful in it being stable,
so occasionally the channel count, and thus values, would get skewed and
the associated lights would change level for enough milliseconds to be
perceived as a flicker. (in this design the ESP 32 was also controlling the
triac phase control of the 60Hz line)

I'd have to go back and see if I made any notes about the detail. But, I
think I also concluded the esp driver code was not written to handle the
situation.

I implemented this many years ago in assembly language on the
ATmega328...I'm not inclined to do it for the ESP32.

It is always possible the project will restart...but, I'm not expecting it
will...

**Comment by pagliara** [comment]
Another workaround is just reverting to version 0.1.0 which doesn't have this issue!
